{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparison TARTAC00 vs. TARTDL00, diastereomeric relationship\n",
    "Based on Respons_2_Norwid, changes start with line 4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": [
     19,
     49,
     66,
     115
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mininum rmsd: 0.0003733837687950025 \n",
      "\n",
      "7\n",
      "aligned\n",
      "C           -0.084          -0.121          -2.952\n",
      "C            0.248           0.383          -1.535\n",
      "C           -0.357          -0.576          -0.494\n",
      "C           -0.025          -0.072           0.923\n",
      "N           -0.589           1.273           1.109\n",
      "O            1.394          -0.021           1.090\n",
      "F           -0.556          -0.913           1.836\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import os, sys\n",
    "if os.path.isfile(os.getenv(\"HOME\") + '/software/ArbAlign/hungarian.cpython-36m-x86_64-linux-gnu.so'):\n",
    "    sys.path.append(os.getenv(\"HOME\") + '/software/ArbAlign/')\n",
    "    import hungarian\n",
    "    linear_sum_assignment =  hungarian.lap\n",
    "else:\n",
    "    from scipy.optimize import linear_sum_assignment\n",
    "\n",
    "from scipy.spatial.distance import cdist\n",
    "\n",
    "AXIS_SWAPS       = np.array([[0,1,2],[0,2,1],[1,0,2],[1,2,0],[2,0,1],[2,1,0]])\n",
    "mask_swaps       =           [  1   ,  -1   ,  -1   ,   1   ,   1   ,  -1  ]\n",
    "AXIS_REFLECTIONS = np.array([[1,1,1],[-1,1,1],[1,-1,1],[1,1,-1],[-1,-1,1],[-1,1,-1],[1,-1,-1],[-1,-1,-1]])\n",
    "mask_reflections =           [  1   ,   -1   ,   -1   ,  -1    ,     1   ,    1    ,    1    ,    -1   ]\n",
    "_rmsd = lambda V, W: np.sqrt(sum([(v[i]-w[i])**2.0 for i in range(len(V[0])) for v, w in zip(V, W)])/len(V))\n",
    "centroid = lambda X: X.mean(axis=0)\n",
    "\n",
    "# Extract element symbols and atomic coordinates from PDB string or filename\n",
    "def parse_pdb(pdb, noHydrogens=True):\n",
    "    \"\"\"\n",
    "    Reads an pdf file into lists\n",
    "    filename - name of xyz file\n",
    "    noHydrogens - if true, hydrogens are ignored; if not, hydrogens are included\n",
    "\n",
    "    Returns a tuple (a, b)\n",
    "    where a is a list of coordinate labels and b is a set of coordinates\n",
    "    (i.e) a = [\"O\", \"H\", \"H\"], b = [[x0,y0,z0],[x1,y1,z1],[x2,y2,z2]]\n",
    "    \"\"\"\n",
    "    pdb = pdb.split('\\n')\n",
    "    if len(pdb) == 1:\n",
    "        with open(pdb[0], 'r') as fr:\n",
    "            pdb = fr.readlines()\n",
    "    labels = []\n",
    "    coords = []\n",
    "    for i in [i for i in pdb if i.find('HETATM') == 0 or i.find('ATOM') == 0 ]:\n",
    "        # ATOM      1  OXT GLY A   3       2.261  -2.612   1.127  1.00  0.00           O\n",
    "        # HETATM    1  C1  UNL     1      -7.227   0.602  -0.445  1.00  0.00           C\n",
    "        ElementSymbol = i[76:78].strip().upper()\n",
    "        #print(noHydrogens)\n",
    "        if noHydrogens and ElementSymbol == \"H\":\n",
    "            continue\n",
    "        else:\n",
    "            labels.append(ElementSymbol)\n",
    "            coords.append([i[30:38],i[38:46],i[46:54]])\n",
    "    #print(unsorted_labels, np.array(unsorted_coords, dtype=float), NA )\n",
    "    return np.asarray(labels), np.array(coords, dtype=float)\n",
    "\n",
    "# Rotate the coordinates of probe molecule according to those of reference molecule\n",
    "def kabsch_rmsd(prb, ref):\n",
    "    # Computation of the covariance matrix\n",
    "    C = np.dot(np.transpose(prb), ref)\n",
    "\n",
    "    # Computation of the optimal rotation matrix\n",
    "    V, S, W = np.linalg.svd(C)\n",
    "\n",
    "    # Ensure a right-handed coordinate system\n",
    "    if (np.linalg.det(V) * np.linalg.det(W)) < 0.0:\n",
    "        S[-1] = -S[-1]\n",
    "        V[:, -1] = -V[:, -1]\n",
    "\n",
    "    # Create Rotation matrix U\n",
    "    U = np.dot(V, W)\n",
    "    prb = np.dot(prb, U)\n",
    "    return _rmsd(prb, ref), prb\n",
    "\n",
    "def reorder_hungarian(p_atoms, q_atoms, p_coord, q_coord):\n",
    "    \"\"\"\n",
    "    Re-orders the input atom list and xyz coordinates using the Hungarian\n",
    "    method (using optimized column results)\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    p_atoms : array\n",
    "        (N,1) matrix, where N is points holding the atoms' names\n",
    "    p_atoms : array\n",
    "        (N,1) matrix, where N is points holding the atoms' names\n",
    "    p_coord : array\n",
    "        (N,D) matrix, where N is points and D is dimension\n",
    "    q_coord : array\n",
    "        (N,D) matrix, where N is points and D is dimension\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    view_reorder : array\n",
    "             (N,1) matrix, reordered indexes of atom alignment based on the\n",
    "             coordinates of the atoms\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # Find unique atoms\n",
    "    unique_atoms = np.unique(p_atoms)\n",
    "\n",
    "    # generate full view from q shape to fill in atom view on the fly\n",
    "    view_reorder = np.zeros(q_atoms.shape, dtype=int)\n",
    "    view_reorder -= 1\n",
    "    for atom in unique_atoms:\n",
    "        p_atom_idx, = np.where(p_atoms == atom)\n",
    "        q_atom_idx, = np.where(q_atoms == atom)\n",
    "\n",
    "        A_coord = p_coord[p_atom_idx]\n",
    "        B_coord = q_coord[q_atom_idx]\n",
    "\n",
    "        # should be kabasch here i think\n",
    "        distances = cdist(A_coord, B_coord, 'euclidean')\n",
    "\n",
    "        # Perform Hungarian analysis on distance matrix between atoms of 1st\n",
    "        # structure and trial structure\n",
    "        _, view = linear_sum_assignment(distances)\n",
    "\n",
    "        view_reorder[p_atom_idx] = q_atom_idx[view]\n",
    "\n",
    "    return view_reorder\n",
    "\n",
    "# Description of the rotated molecule in XYZ format\n",
    "def get_xyz(atoms, V, title=\"\", decimals=8):\n",
    "    \"\"\"\n",
    "    Print coordinates V with corresponding atoms to stdout in XYZ format.\n",
    "    Parameters\n",
    "    ----------\n",
    "    atoms : list\n",
    "        List of atomic types\n",
    "    V : array\n",
    "        (N,3) matrix of atomic coordinates\n",
    "    title : string (optional)\n",
    "        Title of molecule\n",
    "    decimals : int (optional)\n",
    "        number of decimals for the coordinates\n",
    "\n",
    "    Return\n",
    "    ------\n",
    "    output : str\n",
    "        Molecule in XYZ format\n",
    "\n",
    "    \"\"\"\n",
    "    N, D = V.shape\n",
    "\n",
    "    fmt = \"{:2s}\" + (\" {:15.\"+str(decimals)+\"f}\")*3\n",
    "\n",
    "    out = list()\n",
    "    out += [str(N)]\n",
    "    out += [title]\n",
    "\n",
    "    for i in range(N):\n",
    "        atom = atoms[i]\n",
    "        atom = atom[0].upper() + atom[1:]\n",
    "        out += [fmt.format(atom, V[i, 0], V[i, 1], V[i, 2])]\n",
    "\n",
    "    return \"\\n\".join(out)\n",
    "\n",
    "def _align(ref_symbol, prb_symbol, ref_coord, prb_coord, ref_cent,\n",
    "          reorder=False, keep_stereo=True):\n",
    "\n",
    "    min_rmsd   = np.inf\n",
    "    min_review = None\n",
    "    tmp_review = None\n",
    "\n",
    "    for swap, i in zip(AXIS_SWAPS, mask_swaps):\n",
    "        for reflection, j in zip(AXIS_REFLECTIONS, mask_reflections):\n",
    "            if keep_stereo and  i * j == -1: continue # skip enantiomers\n",
    "\n",
    "            # Swap, reflect, and translate the probe molecule\n",
    "            tmp_coord  = prb_coord[:, swap]\n",
    "            tmp_coord  = np.dot(tmp_coord, np.diag(reflection))\n",
    "            tmp_coord -= centroid(tmp_coord)\n",
    "\n",
    "            # Reorder the modifed probe\n",
    "            if reorder:\n",
    "                tmp_review = reorder_hungarian(ref_symbol, prb_symbol, ref_coord, tmp_coord)\n",
    "                tmp_coord = tmp_coord[tmp_review]\n",
    "\n",
    "            # Rotation\n",
    "            this_rmsd, tmp_coord  = kabsch_rmsd(tmp_coord, ref_coord)\n",
    "\n",
    "            if this_rmsd < min_rmsd:\n",
    "                min_rmsd    = this_rmsd\n",
    "                min_review  = tmp_review\n",
    "                min_coord   = tmp_coord\n",
    "                i_j_product = i * j \n",
    "\n",
    "    if reorder:\n",
    "        prb_symbol = prb_symbol[min_review]\n",
    "    prb_coord = min_coord + ref_cent\n",
    "\n",
    "    prb_xyz = get_xyz(prb_symbol, prb_coord, title=\"aligned\", decimals=3)\n",
    "    return min_rmsd, prb_xyz, i_j_product\n",
    "\n",
    "def align(ref_pdb, prb_pdb, noHydrogens=True, reorder=False, keep_stereo=True):\n",
    "    ref_symbol, ref_coord = parse_pdb(ref_pdb, noHydrogens)\n",
    "    prb_symbol, prb_coord = parse_pdb(prb_pdb, noHydrogens)\n",
    "    if len(ref_symbol) != len(prb_symbol): raise\n",
    "    ref_cent = centroid(ref_coord)\n",
    "    prb_cent = centroid(prb_coord)\n",
    "    ref_coord -= ref_cent\n",
    "    prb_coord -= prb_cent\n",
    "    result = _align(ref_symbol, prb_symbol,\n",
    "                    ref_coord, prb_coord, ref_cent,\n",
    "                    reorder, keep_stereo)\n",
    "    return result\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    ref = '''REMARK   1 File created by GaussView 5.0.9\n",
    "HETATM    1  C           0      -0.084  -0.121  -2.952                       C\n",
    "HETATM    2  H           0       0.329  -1.099  -3.088                       H\n",
    "HETATM    3  H           0       0.335   0.547  -3.675                       H\n",
    "HETATM    4  H           0      -1.146  -0.162  -3.076                       H\n",
    "HETATM    5  C           0       0.248   0.383  -1.535                       C\n",
    "HETATM    6  H           0      -0.163   1.362  -1.400                       H\n",
    "HETATM    7  H           0       1.310   0.422  -1.410                       H\n",
    "HETATM    8  C           0      -0.357  -0.576  -0.494                       C\n",
    "HETATM    9  H           0       0.053  -1.555  -0.629                       H\n",
    "HETATM   10  H           0      -1.419  -0.615  -0.619                       H\n",
    "HETATM   11  C           0      -0.025  -0.072   0.923                       C\n",
    "HETATM   12  N           0      -0.589   1.273   1.109                       N\n",
    "HETATM   13  H           0      -1.548   1.197   1.383                       H\n",
    "HETATM   14  H           0      -0.077   1.754   1.821                       H\n",
    "HETATM   15  O           0       1.394  -0.021   1.090                       O\n",
    "HETATM   16  H           0       1.608  -0.064   2.025                       H\n",
    "HETATM   17  F           0      -0.556  -0.913   1.836                       F\n",
    "END '''\n",
    "    prb = '''TITLE       Required\n",
    "REMARK   1 File created by GaussView 5.0.9\n",
    "HETATM    1  C           0     -16.365  -0.185   0.225                       C\n",
    "HETATM    2  H           0     -17.102  -0.670  -0.381                       H\n",
    "HETATM   11  C           0     -14.053  -3.247   0.772                       C\n",
    "HETATM   12  N           0     -13.039  -2.580   1.602                       N\n",
    "HETATM   13  H           0     -13.172  -2.836   2.560                       H\n",
    "HETATM    9  H           0     -16.088  -2.904   0.178                       H\n",
    "HETATM   10  H           0     -15.712  -2.333   1.785                       H\n",
    "HETATM   16  H           0     -13.006  -4.136  -0.641                       H\n",
    "#HETATM   17  F           0     -14.311  -4.478   1.268                       F\n",
    "HETATM   17  F           0     -14.301  -4.478   1.268                       F\n",
    "HETATM   14  H           0     -12.127  -2.862   1.306                       H\n",
    "HETATM   15  O           0     -13.570  -3.362  -0.569                       O\n",
    "HETATM    3  H           0     -16.167   0.791  -0.166                       H\n",
    "HETATM    4  H           0     -16.728  -0.102   1.229                       H\n",
    "HETATM    5  C           0     -15.067  -1.015   0.216                       C\n",
    "HETATM    6  H           0     -14.330  -0.528   0.821                       H\n",
    "HETATM    7  H           0     -14.705  -1.100  -0.787                       H\n",
    "HETATM    8  C           0     -15.351  -2.418   0.781                       C\n",
    "END '''\n",
    "    min_rmsd, prb_xyz, i_j_product = align(ref, prb, noHydrogens=True, reorder=True, keep_stereo=False)\n",
    "    print('mininum rmsd:',min_rmsd,'\\n')\n",
    "    print(prb_xyz)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TARTAC00.pdb as ref (reference molecule)\n",
    "# TARTDL00.pdb as prb (probe molecule)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref = '''CRYST1    7.710    6.000    6.230  90.00 100.10  90.00 P 1 21 1      2\n",
    "TITLE OLEX2 export TARTAC00O-noH.pdb\n",
    "ATOM      1   C1                 3.051   0.095  -4.396  1.00  3.95           C  \n",
    "ATOM      2   C2                 4.461  -0.041  -4.961  1.00  3.95           C  \n",
    "ATOM      3   C3                 2.439   1.364  -5.023  1.00  3.95           C  \n",
    "ATOM      4   C4                 0.989   1.517  -4.603  1.00  3.95           C  \n",
    "ATOM      5   O5                 3.143   0.206  -3.004  1.00  3.95           O  \n",
    "ATOM      6   O6                 4.429  -0.195  -6.268  1.00  3.95           O  \n",
    "ATOM      7   O7                 5.455   0.030  -4.290  1.00  3.95           O  \n",
    "ATOM      8   O8                 3.199   2.504  -4.701  1.00  3.95           O  \n",
    "ATOM      9   O9                 0.307   0.475  -5.036  1.00  3.95           O  \n",
    "ATOM     10  O10                 0.543   2.443  -3.989  1.00  3.95           O    \n",
    "'''\n",
    "prb = '''CRYST1    8.060    9.600    4.850  70.38  97.20 112.47 P -1          2\n",
    "TITLE OLEX2 export, TARTDL00O-noH.pdb\n",
    "ATOM      1   O1                -3.274   2.553   1.873  1.00  3.95           O  \n",
    "ATOM      2   O2                -3.623   4.704   1.964  1.00  3.95           O  \n",
    "ATOM      3   O3                -5.679   4.429   3.655  1.00  3.95           O  \n",
    "ATOM      4   O4                -6.606   3.162   1.325  1.00  3.95           O  \n",
    "ATOM      5   O5                -8.383   2.115   3.015  1.00  3.95           O  \n",
    "ATOM      6   O6                -6.840   1.593   4.523  1.00  3.95           O  \n",
    "ATOM      7   C7                -3.973   3.564   2.193  1.00  3.95           C  \n",
    "ATOM      8   C8                -5.064   3.230   3.061  1.00  3.95           C  \n",
    "ATOM      9   C9                -6.116   2.394   2.330  1.00  3.95           C  \n",
    "ATOM     10  C10                -7.227   2.129   3.381  1.00  3.95           C   \n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test options reorder and keep_stereo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mininum rmsd: 2.2469098354289057 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "min_rmsd, prb_xyz, i_j_product = align(ref, prb, noHydrogens=True, reorder=False, keep_stereo=False)\n",
    "print('mininum rmsd:',min_rmsd,'\\n') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mininum rmsd: 2.379347683339732 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "min_rmsd, prb_xyz, i_j_product = align(ref, prb, noHydrogens=True, reorder=False, keep_stereo=True)\n",
    "print('mininum rmsd:',min_rmsd,'\\n') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rmsd 0.16249943738098105 \n",
      "i_j_product:  1 \n",
      "\n",
      " 10\n",
      "aligned\n",
      "C            3.050           0.080          -4.369\n",
      "C            4.483           0.055          -4.964\n",
      "C            2.363           1.278          -5.027\n",
      "C            1.019           1.431          -4.552\n",
      "O            3.108           0.337          -3.038\n",
      "O            4.572           0.046          -6.281\n",
      "O            5.405          -0.271          -4.248\n",
      "O            3.158           2.502          -4.828\n",
      "O            0.254           0.494          -4.937\n",
      "O            0.605           2.445          -4.027\n"
     ]
    }
   ],
   "source": [
    "min_rmsd, prb_xyz, i_j_product = align(ref, prb, noHydrogens=True, reorder=True, keep_stereo=False)\n",
    "print('rmsd', min_rmsd, '\\ni_j_product: ',i_j_product, '\\n\\n', prb_xyz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rmsd 0.16249943738098105 \n",
      "i_j_product:  1 \n",
      "\n",
      " 10\n",
      "aligned\n",
      "C            3.050           0.080          -4.369\n",
      "C            4.483           0.055          -4.964\n",
      "C            2.363           1.278          -5.027\n",
      "C            1.019           1.431          -4.552\n",
      "O            3.108           0.337          -3.038\n",
      "O            4.572           0.046          -6.281\n",
      "O            5.405          -0.271          -4.248\n",
      "O            3.158           2.502          -4.828\n",
      "O            0.254           0.494          -4.937\n",
      "O            0.605           2.445          -4.027\n"
     ]
    }
   ],
   "source": [
    "min_rmsd, prb_xyz, i_j_product = align(ref, prb, noHydrogens=True, reorder=True, keep_stereo=True)\n",
    "print('rmsd', min_rmsd, '\\ni_j_product: ',i_j_product, '\\n\\n', prb_xyz)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We see the option keep_stereo is responsible for the rmsd difference. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The models are derived from different cif lines of the CSD data base.  The TARTAC datum refers to D-(-)-(2S,3S)-tartaric acid.  TARTDL refers to (2R, 3S)-tartaric acid, wich is both diastereomer and meso-compound with the former.  It was expected that there would no difference in using either --use-reflections or --use-reflections-keep-stereo.\n",
    "\n",
    "However, I notice calculate_rmsd.py from the CLI yields a RMSD of 1.419; then -- with either of the reflection modi -- a RMSD of 0.1625.  Again, none of the three sets of coordinates of the moved motif provided by calculate_rmsd.py on the CLI match the ones provied here.  Quite contrary to expectation, the superposition of the original molecule with the then moved second one /suggests/ a much nicer fit in calculate_rmsd.py default mode, than in each of the others."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
